File: techila/examples/R/Features/function_handle/run_funchandle.r
# Copyright 2010-2013 Techila Technologies Ltd.

# This R-script contains the Local Control Code, which will be used to
# distribute computations to the Techila environment.
#
# A function handle to the 'mcpi_dist' function will be given to the
# funcname parameter, meaning the 'mcpi_dist' function will be
# executed on Workers. The 'loops' parameter will be transferred to
# all Jobs. The peachvector will be used to control the number of Jobs
# in the Project.
#
#
# Usage:
# source("run_funchandle.r")
# result <- run_funchandle(jobs,loops)
# jobs: number of Jobs in the Project
# loops: number of Monte Carlo approximations performed per Job
#
# Example:
# result <- run_funchandle(10,100000)
library(techila)

# This function contains the Worker Code, which will be distributed
# and executed on Workers. The values of the input parameters will be
# received from the parameters defined in the Local Control Code.
mcpi_dist <- function(loops) {
  count <- 0
  for (i in 1:loops) {
    if ((sum(((runif(1) ^ 2)  + (runif(1) ^ 2))) ^ 0.5) < 1) {
      count <- count + 1
    }
  }
  return(count)
}

# This function will distribute create the computational Project by
# using peach.
run_funchandle <- function(jobs,loops) {

  result <- peach(funcname = mcpi_dist,  # Name of the function executed on Workers
                  params = list(loops),   # Input parameters for the executable function
                  peachvector = 1:jobs,   # Length of the peachvecto determines number of Jobs
                  sdkroot = "../../../.." # Location of the techila_settings.ini file
                 )
  # Calculate the approximated value of Pi
  result <- 4 * sum(as.numeric(result)) / (jobs * loops)

  # Display results
  print(c("The approximated value of Pi is:", result))
  result
}

File: techila/examples/python/Features/function_handle/run_funchandle.py
# Copyright 2012-2016 Techila Technologies Ltd.

# This Python contains the Local Control Code, which will be used to
# distribute computations to the Techila environment.
#
# Usage:
# execfile('run_funchandle.py')
# result = run_funchandle(jobcount,loops)
#
# jobcount: number of Jobs in the Project
# loops: number of Monte Carlo approximations performed per Job
#
# Example:
# result = run_funchandle(10, 100000)
import techila

# This function executed on Workers. The values of the input parameters will be
# received from the parameters defined in the Local Control Code.
def mcpi_dist(loops):
    import random
    count = 0
    i = 0
    while i < loops:
        if point_dist() < 1:
            count = count + 1
        i = i + 1

    return count

# This is a subfunction called from mcpi_dist and executed also on Workers.
def point_dist():
    return pow(pow(random.random(), 2) + pow(random.random(), 2), 0.5)

# This function will distribute create the computational Project by
# using peach.
def run_funchandle(jobcount, loops):
    result = techila.peach(funcname = mcpi_dist,    # Execute the mcpi_dist-function on Workers
                           funclist = [point_dist], # Additional function required on Workers
                           params = [loops],        # Input argument to the mcpi_dist-function
                           jobs=jobcount,           # Number of Jobs in the Project
                           )

    # Calculate the approximated value of Pi
    result = 4 * float(sum(result)) / (jobcount * loops)

    # Display results
    print('The approximated value of Pi is:', result)
    return result
